### DISCUSSION TITLE
Proxy port usage and supported configurations

### DISCUSSION THREAD

**Discussion by viniarck** [kytos-ng/telemetry_int#137] [General]
## Section 1 - Problem being solved

- p1) **Business & use case scalability**: whether or not dedicated proxy ports scales well financially and in terms of network capacity planning compared to the feature it enables
- p2) **Code maintenance**: is it worth maintaining proxy port related code to support the business case, how much code and code complexity?

Ultimately, the problem p1) being solved will dictate whether or not proxy ports will still be maintained or not. This issue discussion will focus on p2), which would also help to provide indirect insights and the effort required when deciding on p1).

## Section 2 - Proxy port high level trade off and considerations

- a) It just exist to circumvent a pipeline action limitation in order to allow the sink switch to also add INT metadata
- b) It also enable intra EVCs to be INT monitored
- c) It requires one proxy port per UNI with the current constraints, which might or not be a problem if enough physical (or logical programmable) ports and cables are available on the switch.
- d) Requires significant amount of code to also manage this proxy port state and make sure that INT EVCs paths use it

## Section 3 - Code maintenance in terms of proxy ports

In summary, the following major code functionalities need to exist to support proxy ports (which have already been implemented):

1) topology link_down|link_up events need to handled to converge accordingly
2) topology proxy_port metadata updated|removed needs to be handled accordingly since either the proxy port might no longer exist or it might have changed
3) In memory proxy_port state need to be maintained including the associated unis for validation and to also only allow INT to be enabled if both sink switches have UP proxy ports
3.1) Validation to allow only dedicated proxy ports
3.2) `ProxyPort` abstraction and helpers to represent the UP state and potential related exceptions when the proxy port isn't ready yet
3.3) Keeping track which EVCs are over which ports in order to remove INT flows when falling back to `mef_eline` when a proxy port goes down
4) Auxiliary endpoints and UI to configure proxy ports were implemented to avoid accidental updates where it would disrupt other INT EVCs
5) When an EVC with failover_path gets its current_path or failover_path removed there are extra conditionals to see if the proxy port flow can be removed since it can get overlapped (due to the loop always using the same incoming port and s-vlans of different paths might overlap)

All in all, this represents a large proportion of the code, roughly 25% of `telemetry_int`'s code, and certain parts are also part of the hot path as far as network convergence. Which also implies, more logic to maintain, debug and troubleshoot. 

## Section 4 - Supported configurations

Currently, proxy ports are required on both switch sink A and sink Z to enable INT. This is what we typically call a symmetric proxy port configuration since its symmetry always exist, certain business code logic and flows actions become similar for both directions of a given network path. 

> Asymmetric configurations (which we don't support at the moment) are harder to reason and to maintain. An asymmetric configuration implies also treating each EVC path unidirectionally, which brings even more challenges to derive the direction of the path when getting the flows, and `telemetry_int` shouldn't try to support it until `mef_eline` has a strong business case reason to support it too. 

That said, to keep the code easier to reason, reliable, and maintain with quality we should consider supporting one of these options, option a) being the current one:

- **a)** The current symmetric configuration with proxy ports on both sink switches.
- **b)** No proxy ports at all.
- **c)** Optional choosing a) or b), so the user would specify whether proxy port are required or not.

Option c) is the most flexible, but, it still has the drawbacks of maintaining all the existing code. Plus, it would extra conditionals in the code to derive accordingly when proxy port is required or if it matters for a given EVC. To potentially debug it would be the hardest too, because you'd always need to keep in mind and correlate if the proxy port is involved or not. 

Option b) is the simplest one to maintain, reason, and in terms of reliability:

- Having no proxy ports it implies that all network convergence follows exactly what `mef_eline` does and nothing else extra. So, when debugging events and potential events issues, it's even a smaller surface to deal with.
- The number of network hops and paths would be the same, so correlation with `mef_eline` paths would become simpler to make. So, debugging with SDN trace would also be easier for the network operator to infer what's the expected result of the trace (no need to know which proxy port would be expected). Plus, the confusing case of overlapping with s-vlan on the sink switch external loop wouldn't no longer exist either, that sometimes makes you wonder if the flow got lost or if indeed the overlap should happen.
- It also allows us to remove 25% of the existing code, so fewer tests to maintain and cover, including fewer e2e future test cases that would also need to cover the proxy ports
- `telemetry_int` also following closely and doing nothing more in terms extra path hops, also allows it to be more portable and easier to refactor and evolve. So, whether evolving it as a NApp or even refactor it as feature on `mef_eline`. It still makes sense to keep this as a NApp, and if we keep the events simpler it shouldn't be difficult to maintain.

## Recommendations

As highlighted in Section 1, the decision will be driven by AmLight use cases and network engineering needs. As a core developer, I (Vinicius) recommend option 4b), and completely drop proxy ports if it's acceptable to only have INT visibility from the source to the penultimate sink hop, and not to have INT visibility on intra EVCs. It would become so much easier to maintain, reason and evolve this NApp. Removing the proxy ports also imply not supporting it again in the long-term, but if that decision is hard to make, then we should keep as it is, and only try to revisit this again in the future. Option c) would require even more code to maintain and reason, that option that might interesting and strategic too depending on how we go about it, like only use proxy port minimally in a few EVCs.

**Comment by viniarck** [comment]
What are you guys thoughts, any other considerations or recommendations? Can we drop proxy ports for good? 

cc'ing also @jab1982 and @italovalcy

**Reply by viniarck** [reply]
If positive, I'll continue to augment the doc with how the flows would need to be setup not sending to the loop anymore. But, one step at a time, let's make a decision on the high level options first.

**Comment by jab1982** [comment]
I agree with your suggestion for 4b. However, having no INT at all for intra-switch EVC creates a visibility black hole for the operators. We need to think this carefully.

**Comment by viniarck** [comment]
@jab1982, much appreciated your review. Let's keep thinking about the intra-EVC case and try to settle over the next days, here's a summary of the things to consider and next steps if we can go for option 4b):

#### Proposal next steps considerations

- Go for option 4b) no longer maintain any proxy ports
- Drop support for enabling intra-EVC for now, and only allow enabling INT for inter-EVC. In the future, implement a new feature where it also forward the traffic to a monitoring port (no INT headers though), this is kind of a mirroring feature that can get refined in the future. And/or in the future, only have it when P4 and P4Runtime is supported on Kytos-ng, just so the pipeline design and its maintainability can be done optimally minimizing the efforts involved in the controller and NApps. Would that be acceptable in the meantime, keep that future direction in mind?
- For inter-EVC, multi table pushing and popping INT headers still apply, except on the the sink switch, instead of sending to the external loop (proxy port) when receiving from the NNI, it'll instead for both TCP and UDP: table 0 `send_report`, `goto_table 2|3`, and then on table `2|3` it'll `pop_int` and pop the svlan before forwarding to the UNI. (I'll provide a full example of the flows here for confirmation). 

If we settle on option 4b) then I'll map the rest of the tasks on this epic that needs to be done (mostly covering the bullet points in section 3 of the discussion)

**Reply by viniarck** [reply]
Another great thing of keeping `telemetry_int` in terms of convergence as close to `mef_eline` as possible is that we rely heavily on it, also simplifying more code for the consistency and its maintainability

**Reply by viniarck** [reply]
Example of a sink switch with NNI port 5, proxy port 25-26 and UNI 22 using a single path. This is what we have currently regarding decapsulating from NNI -> UNI:


```
        {
            "flow": {
                "table_id": 0,
                "owner": "telemetry_int",
                "table_group": "evpl",
                "priority": 20100,
                "cookie": 12164322079240733516,
                "idle_timeout": 0,
                "hard_timeout": 0,
                "match": {
                    "in_port": 5,
                    "dl_type": 2048,
                    "dl_vlan": 1,
                    "nw_proto": 6
                },
                "instructions": [
                    {
                        "instruction_type": "apply_actions",
                        "actions": [
                            {
                                "action_type": "add_int_metadata"
                            },
                            {
                                "action_type": "output",
                                "port": 25
                            }
                        ]
                    }
                ]
            },
            "flow_id": "86ad34e0ff51f0d9f90feb7f5b989798",
            "id": "ed2425cc09caf43fa89d0a0d0343ffba",
            "inserted_at": "2024-09-30T14:00:45.109000",
            "state": "installed",
            "switch": "00:00:00:00:00:00:00:06",
            "updated_at": "2024-09-30T14:00:45.334000"
        },
        {
            "flow": {
                "table_id": 0,
                "owner": "telemetry_int",
                "table_group": "evpl",
                "priority": 20100,
                "cookie": 12164322079240733516,
                "idle_timeout": 0,
                "hard_timeout": 0,
                "match": {
                    "in_port": 5,
                    "dl_type": 2048,
                    "dl_vlan": 1,
                    "nw_proto": 17
                },
                "instructions": [
                    {
                        "instruction_type": "apply_actions",
                        "actions": [
                            {
                                "action_type": "add_int_metadata"
                            },
                            {
                                "action_type": "output",
                                "port": 25
                            }
                        ]
                    }
                ]
            },
            "flow_id": "09ccffef66637c2bce00d803d7fee030",
            "id": "68ae8388834827d579f6e2bcfdb7930b",
            "inserted_at": "2024-09-30T14:00:45.109000",
            "state": "installed",
            "switch": "00:00:00:00:00:00:00:06",
            "updated_at": "2024-09-30T14:00:45.334000"
        },
        {
            "flow": {
                "table_id": 0,
                "owner": "telemetry_int",
                "table_group": "evpl",
                "priority": 20000,
                "cookie": 12164322079240733516,
                "idle_timeout": 0,
                "hard_timeout": 0,
                "match": {
                    "in_port": 26,
                    "dl_vlan": 1
                },
                "instructions": [
                    {
                        "instruction_type": "apply_actions",
                        "actions": [
                            {
                                "action_type": "send_report"
                            }
                        ]
                    },
                    {
                        "instruction_type": "goto_table",
                        "table_id": 2
                    }
                ]
            },
            "flow_id": "953ff6e0d12cbadd375199ecb7ad9937",
            "id": "742ea570d415530112f2f2a5c261fed3",
            "inserted_at": "2024-09-30T14:00:45.109000",
            "state": "installed",
            "switch": "00:00:00:00:00:00:00:06",
            "updated_at": "2024-09-30T14:00:45.334000"
        },
        {
            "flow": {
                "table_id": 2,
                "owner": "telemetry_int",
                "table_group": "evpl",
                "priority": 20000,
                "cookie": 12164322079240733516,
                "idle_timeout": 0,
                "hard_timeout": 0,
                "match": {
                    "in_port": 26,
                    "dl_vlan": 1
                },
                "instructions": [
                    {
                        "instruction_type": "apply_actions",
                        "actions": [
                            {
                                "action_type": "pop_int"
                            },
                            {
                                "action_type": "pop_vlan"
                            },
                            {
                                "action_type": "output",
                                "port": 22
                            }
                        ]
                    }
                ]
            },
            "flow_id": "fc6ebd48fe7e6b1f03e22f80df7439b4",
            "id": "7cf282352e881c9eed30ba307b02612b",
            "inserted_at": "2024-09-30T14:00:45.109000",
            "state": "installed",
            "switch": "00:00:00:00:00:00:00:06",
            "updated_at": "2024-09-30T14:00:45.334000"
        }
```

**Reply by viniarck** [reply]
This is what we can have if we were to drop the proxy port. NNI port 5 and UNI 22 using a single path, decapsulating from NNI -> UNI:

```
    {
      "table_id": 0,
      "owner": "telemetry_int",
      "table_group": "evpl",
      "priority": 20100,
      "cookie": 12164322079240733516,
      "idle_timeout": 0,
      "hard_timeout": 0,
      "match": {
        "in_port": 5,
        "dl_type": 2048,
        "dl_vlan": 1,
        "nw_proto": 6
      },
      "instructions": [
        {
          "instruction_type": "apply_actions",
          "actions": [
            {
              "action_type": "send_report"
            }
          ]
        },
        {
          "instruction_type": "goto_table",
          "table_id": 2
        }
      ]
    },
    {
      "table_id": 0,
      "owner": "telemetry_int",
      "table_group": "evpl",
      "priority": 20100,
      "cookie": 12164322079240733516,
      "idle_timeout": 0,
      "hard_timeout": 0,
      "match": {
        "in_port": 5,
        "dl_type": 2048,
        "dl_vlan": 1,
        "nw_proto": 17
      },
      "instructions": [
        {
          "instruction_type": "apply_actions",
          "actions": [
            {
              "action_type": "send_report"
            }
          ]
        },
        {
          "instruction_type": "goto_table",
          "table_id": 2
        }
      ]
    },
    {
      "table_id": 2,
      "owner": "telemetry_int",
      "table_group": "evpl",
      "priority": 20000,
      "cookie": 12164322079240733516,
      "idle_timeout": 0,
      "hard_timeout": 0,
      "match": {
        "in_port": 5,
        "dl_vlan": 1
      },
      "instructions": [
        {
          "instruction_type": "apply_actions",
          "actions": [
            {
              "action_type": "pop_int"
            },
            {
              "action_type": "pop_vlan"
            },
            {
              "action_type": "output",
              "port": 22
            }
          ]
        }
      ]
    }
```

**Comment by viniarck** [comment]
As we spoke on today's meeting: this will be open to feedback by Oct 11.

**Comment by italovalcy** [comment]
Hi @viniarck and @jab1982, great discussion so far! Excellent points you highlighted, @viniarck! 

I understand the main discussion here is regarding the decision to keep the proxy port or not. As Jeronimo mentioned, not having a proxy port would impact visibility for intra-switch EVCs. Currently, 169 out of 444 EVCs are intra-switch, 38%. I understand there is complexity associated with maintaining proxy ports. IMHO, it is a price we must pay to have visibility on the 38% of the EVCs mentioned above. In the future, when we have a better solution for not losing visibility on the intra-switch EVCs without the proxy port, we can re-assess this topic. 

Thus, I believe the best option now would be to keep support for the proxy port and add support for symmetrically doing penultimate hop INT sink.

**Reply by viniarck** [reply]
Good morning, @italovalcy. No problem, let's maintain it then, the decision has to be primarily driven by the business use case, let's keep supporting proxy ports with only their symmetric configuration. Most of the development cost has been paid, and if it's understood the complexity and opportunity to simplify and weighing in and if the conclusion is that it's important for the business let's bite the bullet and continue to maintain. In terms of extra new development, the remaining significant parts are consistency checks and e2e tests (we also have that limitation and validation of exclusive proxy port per UNI but no other concerns in terms of use case have been raised so far), with e2e tests if we cover well proxy ports then it should be doable to maintain it with good quality.

**Reply by viniarck** [reply]
Final question before we close this issue decision: will AmLight in the short/medium term have the proxy ports and cables available for the symmetric configuration deployment once we have e2e test covered too? I'm trying to understand if you guys already have the cables or spare cables already planned for the future deployment. Thanks for the the reviews and your inputs @jab1982 and @italovalcy
